src/github.com/matrix-org/go-neb/services/imgur/imgur.go

```diff
@@ -0,0 +1,270 @@
+	// The API key to use when making HTTP requests to Imgur.
+	ClientSecret string `json:"client_secret"`
+}
+
+// Commands supported:
+//    !imgur some_search_query_without_quotes
+// Responds with a suitable image into the same room as the command.
+func (s *Service) Commands(client *gomatrix.Client) []types.Command {
+	return []types.Command{
+		types.Command{
+			Path: []string{"imgur", "help"},
+			Command: func(roomID, userID string, args []string) (interface{}, error) {
+				return usageMessage(), nil
+			},
+		},
+		types.Command{
+			Path: []string{"imgur"},
+			Command: func(roomID, userID string, args []string) (interface{}, error) {
+				return s.cmdImgurImgSearch(client, roomID, userID, args)
+			},
+		},
+	}
+}
+
+// usageMessage returns a matrix TextMessage representation of the service usage
+func usageMessage() *gomatrix.TextMessage {
+	return &gomatrix.TextMessage{"m.notice",
+		`Usage: !imgur image_search_text`}
+}
+
+func (s *Service) cmdImgurImgSearch(client *gomatrix.Client, roomID, userID string, args []string) (interface{}, error) {
+
+	if len(args) < 1 {
+		return usageMessage(), nil
+	}
+
+	// Get the query text to search for.
+	querySentence := strings.Join(args, " ")
+
+	searchResultImage, searchResultAlbum, err := s.text2imgImgur(querySentence)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Image returned
+	if searchResultImage != nil {
+		var imgURL = searchResultImage.Link
+		if imgURL == "" {
+			return gomatrix.TextMessage{
+				MsgType: "m.notice",
+				Body:    "No image found!",
+			}, nil
+		}
+
+		// FIXME -- Sometimes upload fails with a cryptic error - "msg=Upload request failed code=400"
+		// log.Printf("Uploading image at: %s", imgURL)
+		resUpload, err := client.UploadLink(imgURL)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to upload Imgur image (%s) to matrix: %s", imgURL, err.Error())
+		}
+
+		return gomatrix.ImageMessage{
+			MsgType: "m.image",
+			Body:    querySentence,
+			URL:     resUpload.ContentURI,
+			Info: gomatrix.ImageInfo{
+				Height:   uint(searchResultImage.Height),
+				Width:    uint(searchResultImage.Width),
+				Mimetype: searchResultImage.Type,
+			},
+		}, nil
+	} else if searchResultAlbum != nil {
+		return gomatrix.TextMessage{
+			MsgType: "m.notice",
+			Body:    "Search returned an album - Not currently supported",
+		}, nil
+	} else {
+		return gomatrix.TextMessage{
+			MsgType: "m.notice",
+			Body:    "No image found!",
+		}, nil
+	}
+}
+
+// text2imgImgur returns info about an image or an album
+func (s *Service) text2imgImgur(query string) (*imgurGalleryImage, *imgurGalleryAlbum, error) {
+	log.Info("Searching Imgur for an image of a ", query)
+
+	query = url.QueryEscape(query)
+	var base = "https://api.imgur.com/3/gallery/search"
+	var sort = "time"  // time | viral | top
+	var window = "all" // day | week | month | year | all
+	var page = 1
+	var urlString = fmt.Sprintf("%s/%s/%s/%d?q=%s", base, sort, window, page, query)
```

Comment:
I like the `sort`, `window` and `page` vars which are used to add more information about how the URL is constructed. I don't like the `base`, which adds nothing. Can you please inline `base`?